crates/data_components/src/lib.rs

```diff
@@ -0,0 +1,49 @@
+/*
+Copyright 2024 The Spice.ai OSS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#![allow(clippy::missing_errors_doc)]
+use std::{error::Error, sync::Arc};
+
+use async_trait::async_trait;
+use datafusion::{common::OwnedTableReference, datasource::TableProvider};
+
+pub mod arrow;
+pub mod databricks;
+
+#[async_trait]
+pub trait Read {
+    async fn table_provider(
+        &self,
+        table_reference: OwnedTableReference,
+    ) -> Result<Arc<dyn TableProvider + 'static>, Box<dyn Error + Send + Sync>>;
```

Comment:
This is a great idea - I originally had this as a concrete error type that all providers would have to adhere to - which is inflexible, because you basically have to aggregate all the possibles errors from each provider into one error type, or you have one super-generic error in the concrete instead which doesn't really help. However using associated types is great, and something I didn't think about before.

The only thing I can think of is I'm not sure if this would break the ability to make this a trait object, which as I mentioned above we want to keep. But I'll test it out.

Reply:
So it can remain as a trait object, but then we would need to specify the associated type whenever working with the trait object itself, which partially defeats the point. I'll keep as is for now.